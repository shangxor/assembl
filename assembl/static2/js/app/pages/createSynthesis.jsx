// @flow
import React from 'react';
import { FormGroup, Row, Col } from 'react-bootstrap';
import Section from '../components/common/section';
import FormControlWithLabel from '../components/common/formControlWithLabel';

const CreateSynthesis = () => (
  <div className="administration max-container create-synthesis">
    <Section title="debate.syntheses.createNewSynthesis" translate>
      <Row>
        <Col xs={8}>
          <div className="admin-box">
            <FormControlWithLabel label="Titre de la synthÃ¨se" required type="text" value="" onChange={() => {}} />
          </div>
        </Col>e
      </Row>
    </Section>
  </div>
);

export default CreateSynthesis;