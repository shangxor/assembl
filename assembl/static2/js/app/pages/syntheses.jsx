// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Link } from 'react-router';
import { Localize, Translate } from 'react-redux-i18n';

import { get } from '../utils/routeMap';
import { browserHistory } from '../router';
import { CLASS_NAME_GENERATOR } from '../utils/cardList';
import Section from '../components/common/section';
import Card from '../components/common/card';
import CardList from '../components/common/cardList';
import SynthesesQuery from '../graphql/SynthesesQuery.graphql';

import manageErrorAndLoading from '../components/common/manageErrorAndLoading';

type SynthesesProps = {
  syntheses: Array<Object>,
  slug: string,
  hasSyntheses: boolean
};

export class DumbSyntheses extends React.Component<SynthesesProps> {
  componentDidMount() {
    const { syntheses, slug } = this.props;
    if (syntheses && syntheses.length === 1) {
      const firstSynthesis = syntheses[0];
      browserHistory.push(`${get('synthesis', { synthesisId: firstSynthesis.post.id, slug: slug })}`);
    }
  }

  render() {
    const { syntheses, slug, hasSyntheses } = this.props;
    return (
      <Section title="debate.syntheses.summary" translate>
        <div className="center create-synthesis">
          <Link className="button-submit button-dark">
            <Translate value="debate.syntheses.createNewSynthesis" />
          </Link>
        </div>
        {!hasSyntheses ? (
          <h2 className="dark-title-2 margin-left-xxl">
            <Translate value="synthesis.noSynthesisYet" />
          </h2>
        ) : (
          <CardList
            data={syntheses}
            classNameGenerator={CLASS_NAME_GENERATOR.default}
            itemClassName="theme"
            CardItem={itemData => (
              <Card imgUrl={itemData.img ? itemData.img.externalUrl : ''} className="synthesis-preview">
                <Link className="content-box" to={`${get('synthesis', { synthesisId: itemData.post.id, slug: slug })}`}>
                  <div className="title-container center">
                    <h3 className="light-title-3">{itemData.subject}</h3>
                    <h4 className="light-title-4">
                      <Localize value={itemData.creationDate} dateFormat="date.format2" />
                    </h4>
                  </div>
                </Link>
              </Card>
            )}
          />
        )}
      </Section>
    );
  }
}

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

export default compose(
  connect(mapStateToProps),
  graphql(SynthesesQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        syntheses: data.syntheses,
        hasSyntheses: data && data.syntheses ? data.syntheses.length > 0 : false
      };
    }
  }),
  manageErrorAndLoading({ displayLoader: true })
)(DumbSyntheses);