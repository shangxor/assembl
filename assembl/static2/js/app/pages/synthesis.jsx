// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Col, Grid, Row } from 'react-bootstrap';
import debounce from 'lodash/debounce';

import { renderSynthesisBody } from '../utils/linkify';
import Header from '../components/common/header';
import Section from '../components/common/section';
import SynthesisQuery from '../graphql/SynthesisQuery.graphql';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import IdeaSynthesisTree from '../components/synthesis/IdeaSynthesisTree';
import { getChildren, getPartialTree } from '../utils/tree';
import SideMenu from '../components/synthesis/sideMenu';
import { getDomElementOffset } from '../utils/globalFunctions';
import type { Synthesis } from '../components/synthesis/types.flow';

type SynthesisProps = {
  synthesis: Synthesis,
  routeParams: { slug: string, synthesisId: string },
  synthesisPostId: string
};

type SynthesisState = {
  introBlock: ?HTMLElement,
  conclusionBlock: ?HTMLElement,
  bodyBlock: ?HTMLElement,
  sideMenuNode: ?HTMLElement,
  sideMenuIsHidden: boolean,
  ideaOnScroll?: string,
  sideMenuHeight: number,
  sideMenuOverflow: string,
  setSideMenuHeight: string
};

const sideMenuTopPercentage = 15;

export class DumbSynthesis extends React.Component<SynthesisProps, SynthesisState> {
  constructor(props: SynthesisProps) {
    super(props);
    this.state = {
      introBlock: null,
      conclusionBlock: null,
      bodyBlock: null,
      sideMenuNode: null,
      sideMenuIsHidden: true,
      sideMenuHeight: 0,
      sideMenuOverflow: 'auto',
      setSideMenuHeight: '100%'
    };
  }

  componentWillMount() {
    window.addEventListener('scroll', this.updateSideMenuVisibility);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.updateSideMenuVisibility);
  }

  updateIntroBlock = (node: HTMLElement) => {
    this.setState({ introBlock: node });
  };

  updateConclusionBlock = (node: HTMLElement) => {
    this.setState({ conclusionBlock: node });
  };

  updateBodyBlock = (node: HTMLElement) => {
    this.setState({ bodyBlock: node });
  };

  updateSideMenuNode = (node: HTMLElement) => {
    this.setState({ sideMenuNode: node });
  };

  updateSideMenuVisibility = debounce(() => {
    const { introBlock, conclusionBlock, sideMenuNode } = this.state;
    if (introBlock && conclusionBlock) {
      const introNodeTop = getDomElementOffset(introBlock).top;
      const introNodeHeight = introBlock.getBoundingClientRect().height;
      const sideMenuHeight = sideMenuNode && sideMenuNode.getBoundingClientRect().height;
      if (sideMenuHeight && sideMenuHeight > 0) {
        this.setState({ sideMenuHeight: sideMenuHeight });
      }
      const conclusionBlockTopOffset = getDomElementOffset(conclusionBlock).top;
      const firstIdeaTopOffset = introNodeTop + introNodeHeight;
      const scroll = window.pageYOffset;
      const { innerHeight, innerWidth } = window;
      const sideMenuTopOffset = innerHeight * sideMenuTopPercentage / 100;
      // this value is the gap between the top of sideMenu and the top of the window set by the css 'top' property
      if (sideMenuHeight && sideMenuHeight + sideMenuTopOffset > innerHeight) {
        this.setState({ sideMenuOverflow: 'scroll', setSideMenuHeight: '80%' });
      }

      const hasScrollReachedSynthesis = scroll + 60 > firstIdeaTopOffset;
      // 60 corresponds to the gap between the first Idea and the introduction block (synthesis-tree's padding)
      const isBottomReached = scroll + sideMenuTopOffset + this.state.sideMenuHeight + 90 >= conclusionBlockTopOffset;
      // 90 corresponds to the gap between the top of the conclusion block and the top of its div (padding)

      if (isBottomReached || !hasScrollReachedSynthesis || innerWidth < 1000) {
        this.setState({ sideMenuIsHidden: true });
      } else {
        this.setState({ sideMenuIsHidden: false });
      }
    }
  }, 100);

  handleScrollForMenu = (node: HTMLElement) => {
    this.setState({ ideaOnScroll: node.id });
  };

  render() {
    const { synthesis, routeParams, synthesisPostId } = this.props;
    const { sideMenuIsHidden, ideaOnScroll, sideMenuOverflow, setSideMenuHeight } = this.state;
    const { introduction, conclusion, body, ideas, subject } = synthesis;
    const sortedIdeas = [...ideas].sort((a, b) => {
      if (a.live.order < b.live.order) {
        return -1;
      }
      if (a.live.order > b.live.order) {
        return 1;
      }
      return 0;
    });
    const { roots, descendants } = getPartialTree(sortedIdeas);
    const hasSiblings = roots.length > 1;
    return (
      <div className="synthesis-page">
        <div className="background-light">
          <Header title={subject} imgUrl={synthesis.img ? synthesis.img.externalUrl : ''} type="synthesis" />
          <Grid fluid>
            {introduction && (
              <Section title="introduction" translate className="synthesis-block" innerRef={this.updateIntroBlock}>
                <Row>
                  <Col mdOffset={3} md={8} smOffset={1} sm={10}>
                    <div dangerouslySetInnerHTML={{ __html: introduction }} />
                  </Col>
                </Row>
              </Section>
            )}
            {body && (
              <Section title="" className="synthesis-block" innerRef={this.updateBodyBlock}>
                <Row>
                  <Col mdOffset={3} md={8} smOffset={1} sm={10}>
                    {renderSynthesisBody(body)}
                  </Col>
                </Row>
              </Section>
            )}
            <Row className="background-grey synthesis-tree">
              <Col
                md={3}
                className="affix"
                style={{
                  top: `${sideMenuTopPercentage} %`,
                  overflow: sideMenuOverflow,
                  height: setSideMenuHeight
                }}
              >
                <SideMenu
                  rootIdeas={roots}
                  descendants={descendants}
                  synthesisPostId={synthesisPostId}
                  ideaOnScroll={ideaOnScroll}
                  innerRef={this.updateSideMenuNode}
                  show={!sideMenuIsHidden}
                />
              </Col>
              <Col mdOffset={3} md={8} sm={11}>
                {roots.map((rootIdea, index) => (
                  <IdeaSynthesisTree
                    hasSiblings={hasSiblings}
                    key={rootIdea.id}
                    rootIdea={rootIdea}
                    index={index + 1}
                    parents={[]}
                    subIdeas={getChildren(rootIdea, descendants)}
                    slug={routeParams.slug}
                    handleScrollForMenu={this.handleScrollForMenu}
                  />
                ))}
              </Col>
            </Row>
            {conclusion && (
              <Section
                title="conclusion"
                translate
                className="synthesis-block"
                id="conclusion"
                innerRef={this.updateConclusionBlock}
              >
                <Row>
                  <Col mdOffset={3} md={8} smOffset={1} sm={10}>
                    <div dangerouslySetInnerHTML={{ __html: conclusion }} />
                  </Col>
                </Row>
              </Section>
            )}
          </Grid>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  lang: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(SynthesisQuery, {
    options: props => ({
      variables: { id: props.params.synthesisId, lang: props.lang }
    }),
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        synthesis: data.synthesisPost.publishesSynthesis,
        synthesisPostId: data.synthesisPost.id
      };
    }
  }),
  manageErrorAndLoading({ displayLoader: true })
)(DumbSynthesis);