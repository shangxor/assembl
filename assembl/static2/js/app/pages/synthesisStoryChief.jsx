// @flow
import React from 'react';
import { connect } from 'react-redux';
import Section from '../components/common/section';

type Props = {
  title: string,
  url: string
};

class Synthesis extends React.Component<Props> {
  handleLoad = () => {};

  render() {
    const { title, url } = this.props;
    return (
      <Section title={title}>
        <div className="synthesis">
          <iframe
            ref={(iframe) => {
              if (iframe) {
                this.iframe = iframe;
              }
            }}
            src={url}
            title="storychief-iframe"
            frameBorder="0"
            className="synthesis-iframe"
          />
        </div>
      </Section>
    );
  }
}

const mockProps = () => ({
  title: 'Un bon titre de synth√®se',
  url: 'https://bluenovev3.storychief.io/culture-manager-phase-2'
});

export default connect(mockProps)(Synthesis);