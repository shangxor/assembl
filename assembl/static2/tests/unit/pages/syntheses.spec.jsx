import React from 'react';
import renderer from 'react-test-renderer';

import { DumbSyntheses } from '../../../js/app/pages/syntheses';

describe('Syntheses component', () => {
  xit('should match Syntheses snapshot', () => {
    const props = {
      hasSyntheses: true,
      syntheses: [
        {
          id: 'fooId',
          subject: 'Foo',
          img: { externalUrl: 'http://foo.com/bar' },
          creationDate: '2017-02-10T09:15:20.707854+00:00',
          post: { id: 'someId' }
        },
        {
          id: 'barId',
          subject: 'Bar',
          img: { externalUrl: 'http://foo.com/bar' },
          creationDate: '2017-02-10T09:15:20.707854+00:00',
          post: { id: 'someId' }
        }
      ],
      slug: 'fooslug'
    };
    const rendered = renderer.create(<DumbSyntheses {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
  it('should match empty Syntheses snapshot', () => {
    const props = { hasSyntheses: false, syntheses: [] };
    const rendered = renderer.create(<DumbSyntheses {...props} />).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});