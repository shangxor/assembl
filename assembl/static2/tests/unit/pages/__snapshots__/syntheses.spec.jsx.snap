// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Syntheses component should match Syntheses snapshot 1`] = `
<section
  className="themes-section"
>
  <div
    className="max-container"
  >
    <div
      className="title-section"
    >
      <div>
        <div
          className="title-hyphen"
        >
           
        </div>
        <h1
          className="section-title section-title1 dark-title-1"
        >
          <span>
            Syntheses summary
          </span>
        </h1>
      </div>
    </div>
    <div
      className="content-section margin-l"
    >
      <div
        className="center create-synthesis"
      >
        <a
          className="button-submit button-dark"
          onClick={[Function]}
          style={Object {}}
        >
          <span>
            Create a new synthesis
          </span>
        </a>
      </div>
      <div
        className="no-margin row"
      >
        <div
          className="theme clear theme-first-row-0 col-md-3 col-sm-6 col-xs-12"
        >
          <div
            className="illustration-box synthesis-preview"
          >
            <div
              className="image-box"
              style={
                Object {
                  "backgroundImage": "url(http://foo.com/bar)",
                }
              }
            />
            <a
              className="content-box"
              onClick={[Function]}
              style={Object {}}
            >
              <div
                className="title-container center"
              >
                <h3
                  className="light-title-3"
                >
                  Foo
                </h3>
                <h4
                  className="light-title-4"
                >
                  <span>
                    2017-02-10
                  </span>
                </h4>
              </div>
            </a>
            <div
              className="color-box"
            >
               
            </div>
            <div
              className="box-hyphen"
            >
               
            </div>
            <div
              className="box-hyphen rotate-hyphen"
            >
               
            </div>
          </div>
        </div>
        <div
          className="theme theme-first-row-1 col-md-3 col-sm-6 col-xs-12"
        >
          <div
            className="illustration-box synthesis-preview"
          >
            <div
              className="image-box"
              style={
                Object {
                  "backgroundImage": "url(http://foo.com/bar)",
                }
              }
            />
            <a
              className="content-box"
              onClick={[Function]}
              style={Object {}}
            >
              <div
                className="title-container center"
              >
                <h3
                  className="light-title-3"
                >
                  Bar
                </h3>
                <h4
                  className="light-title-4"
                >
                  <span>
                    2017-02-10
                  </span>
                </h4>
              </div>
            </a>
            <div
              className="color-box"
            >
               
            </div>
            <div
              className="box-hyphen"
            >
               
            </div>
            <div
              className="box-hyphen rotate-hyphen"
            >
               
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`Syntheses component should match empty Syntheses snapshot 1`] = `
<section
  className="themes-section"
>
  <div
    className="max-container"
  >
    <div
      className="title-section"
    >
      <div>
        <div
          className="title-hyphen"
        >
           
        </div>
        <h1
          className="section-title section-title1 dark-title-1"
        >
          <span>
            Syntheses summary
          </span>
        </h1>
      </div>
    </div>
    <div
      className="content-section margin-l"
    >
      <div
        className="center create-synthesis"
      >
        <a
          className="button-submit button-dark"
          onClick={[Function]}
          style={Object {}}
        >
          <span>
            Create a new synthesis
          </span>
        </a>
      </div>
      <h2
        className="dark-title-2 margin-left-xxl"
      >
        <span>
          There is no synthesis available yet.
        </span>
      </h2>
    </div>
  </div>
</section>
`;
